lib/action.ml

```diff
@@ -182,30 +222,84 @@ module Action (Github_api : Api.Github) (Slack_api : Api.Slack) = struct
     | false ->
     match req with
     | Github.Push n ->
-      partition_push cfg n |> List.map ~f:(fun (channel, n) -> generate_push_notification n channel) |> Lwt.return
-    | Pull_request n -> partition_pr cfg n |> List.map ~f:(generate_pull_request_notification n) |> Lwt.return
-    | PR_review n -> partition_pr_review cfg n |> List.map ~f:(generate_pr_review_notification n) |> Lwt.return
+      partition_push cfg n
+      |> List.map ~f:(fun (channel, n) ->
+           generate_stateful_notification ctx (generate_push_notification_content n) req channel
+         )
+      |> Lwt.return
+    | Pull_request n ->
+      partition_pr cfg n
+      |> List.map ~f:(fun channel ->
+           generate_stateful_notification ctx (generate_pull_request_notification_content n) req channel
+         )
+      |> Lwt.return
+    | PR_review n ->
+      partition_pr_review cfg n
+      |> List.map ~f:(fun channel ->
+           generate_stateful_notification ctx (generate_pr_review_notification_content n) req channel
+         )
+      |> Lwt.return
     | PR_review_comment n ->
-      partition_pr_review_comment cfg n |> List.map ~f:(generate_pr_review_comment_notification n) |> Lwt.return
-    | Issue n -> partition_issue cfg n |> List.map ~f:(generate_issue_notification n) |> Lwt.return
+      partition_pr_review_comment cfg n
+      |> List.map ~f:(fun channel ->
+           generate_stateful_notification ctx (generate_pr_review_comment_notification_content n) req channel
+         )
+      |> Lwt.return
+    | Issue n ->
+      partition_issue cfg n
+      |> List.map ~f:(fun channel ->
+           generate_stateful_notification ctx (generate_issue_notification_content n) req channel
+         )
+      |> Lwt.return
     | Issue_comment n ->
-      partition_issue_comment cfg n |> List.map ~f:(generate_issue_comment_notification n) |> Lwt.return
+      partition_issue_comment cfg n
+      |> List.map ~f:(fun channel ->
+           (* not sure why but when I remove "fun channel" it evaluates the [generate_issue_comment_notification_content n] before the partition so I got errors but when I add this, it's completely fine, probably some compiler optimization? *)
```

Comment:
@yasunariw @Khady any idea why? from what I understand, even without making it an explicit function, the partial evaluation of `f` should happen after the list generation right?